docs(cn): translate content/docs/lifting-state-up.md into Chinese (#90)
reactjs/zh-hans.reactjs.org: content/docs/lifting-state-up.md

---
id: lifting-state-up

> title: 状态提升

Reviewer: state 感觉不需要翻译
```suggestion
title: 提升 State
```

Author: 这个是因为nav里的一模一样的标题被翻译成了英文我为了保持统一所以翻译了

  - "docs/flux-todo-list.html"
---

> 

Reviewer: ditto, state 应该不用翻译

通常，多个组件需要反映相同的变化数据，这时我们建议将共享状态提升到最近的共同父组件中去。让我们看看它是如何运作的。

在本节中，我们将创建一个用于计算水在给定温度下是否会沸腾的温度计算器。

> 我们将从一个名为 `BoilingVerdict` 的组件开始，它接受 `celsius` 温度作为一个属性，并且打印出这个温度是否足以让水沸腾。

Reviewer: ```suggestion
我们将从一个名为 `BoilingVerdict` 的组件开始，它接受 `celsius` 温度作为一个 prop，并且打印出这个温度是否足以让水沸腾。
```

Author: glossary中不翻译的是props，这里的prop应该是property的简写，所以就翻译了，这个还需要大家讨论一下

Reviewer 2: prop 和 props 均暂时保持不译。

    // ...  
```

然而，我们希望这两个输入框中的数值能够保持同步。当我们更新摄氏度输入框的内容时，华氏度输入框应当显示转换后的华氏温度，反之亦然。

> 在 React 中，共享 state 是通过将多个组件中需要共享的 state 向上移动到它们的最近共同父组件来实现的。这就是所谓的“状态提升”。接下来，我们将 `TemperatureInput` 组件中的 state 移动至 `Calculator` 组件中去。

Reviewer: 同上 state 不用翻译

Author: 这个位置是为了与导航中的标题翻译保持同步，与上面的情况相同

    // ...
```

> 我们知道 [props 是只读的](/docs/components-and-props.html#props-are-read-only)。当 `temperature` 在 `TemperatureInput` 组件本地的 state 中时，组件只能通过调用 `this.setState()` 去修改它。然而现在 `temperature` 来源于父组件并以属性的方式传入，`TemperatureInput` 组件也就对它已经没有任何的控制权了。

Reviewer: ```suggestion
我们知道 [props 是只读的](/docs/components-and-props.html#props-are-read-only)。当 `temperature` 在 `TemperatureInput` 组件本地的 state 中时，组件只能通过调用 `this.setState()` 去修改它。然而现在 `temperature` 来源于父组件并以 prop 传入，`TemperatureInput` 组件也就对它已经没有任何的控制权了。
```

Reviewer 2: 「然而现在 `temperature` 作为父组件传入的 prop，」

Author: > 「然而现在 `temperature` 作为父组件传入的 prop，`TemperatureInput` 组件便无法控制它了。」
感觉这么翻有点生硬了

Reviewer 3: 1. 「当 `temperature` 在 `TemperatureInput` 组件本地的 state 中时」 => 「当 `temperature` 存在与组件 `TemperatureInput` 的 state 中时」
2. 然而 `temperature` 是父组件传入的 prop，`TemperatureInput` 组件便失去了对它的控制权

Author: > 1. 「当 `temperature` 在 `TemperatureInput` 组件本地的 state 中时」 => 「当 `temperature` 存在与组件 `TemperatureInput` 的 state 中时」
> 2. 然而 `temperature` 是父组件传入的 prop，`TemperatureInput` 组件便失去了对它的控制权

第一点没问题，针对第二点，我翻译成：
2\. 然而现在， `temperature` 是由父组件传入的 prop，`TemperatureInput` 组件便失去了对它的控制权
这样与上句可以产生一个前后变化的对比，读起来更顺畅一些。

}
```

> 如果我们之后修改华氏度的输入框中的内容为212时，`Calculator` 组件中的 state 将会是：

Reviewer: ```suggestion
如果我们之后修改华氏度的输入框中的内容为 212 时，`Calculator` 组件中的 state 将会是：
```

}
```

[**在 CodePen 上尝试**](https://codepen.io/gaearon/pen/WZpxpz?editors=0010)

现在无论你编辑哪个输入框中的内容，`Calculator` 组件中的 `this.state.temperature` 和 `this.state.scale` 均会被更新。其中一个输入框保留用户的输入并取值，另一个输入框始终基于这个值显示转换后的结果。

让我们来重新梳理一下当你对输入框内容进行编辑时会发生些什么：

* React 在 DOM 的 `<input>` 上调用被指定为是 `onChange` 的函数。在本例中，它是 `TemperatureInput` 组件的 `handleChange` 方法。
* `TemperatureInput` 组件的 `handleChange` 方法使用新输入的值调用 `this.props.onTemperatureChange()`。这个组件的 props 包括 `onTemperatureChange` 在内，都是由它的父组件 `Calculator` 提供的。

> * 当 `Calculator` 组件先前渲染后，它用于摄氏度输入的子组件 `TemperatureInput` 中的 `onTemperatureChange` 被指定为了 `Calculator` 组件中的 `handleCelsiusChange` 方法，并且，其用作华氏度输入的子组件 `TemperatureInput` 中的 `onTemperatureChange` 被指定为了 `Calculator` 组件中的 `handleFahrenheitChange` 方法。因此，哪个输入框被编辑将决定 `Calculator` 组件上哪个对应的方法被调用。

Reviewer 3: ```suggestion
* 起初渲染时，用于摄氏度输入的子组件 `TemperatureInput` 中 `onTemperatureChange` 方法为 `Calculator` 组件中的 `handleCelsiusChange` 方法，而，用于华氏度输入的子组件 `TemperatureInput` 中的 `onTemperatureChange` 方法为 `Calculator` 组件中的 `handleFahrenheitChange` 方法。因此，无论哪个输入框被编辑都会调用 `Calculator` 组件中对应的方法。
```

* 在这些方法内部，`Calculator` 组件通过使用新的输入值与当前输入框对应的温度计量单位来调用 `this.setState()` 进而请求 React 重新渲染自己本身。
* React 调用 `Calculator` 组件的 `render` 方法得到组件的 UI 呈现。温度转换在这时进行，两个输入框中的数值通过当前输入温度和其计量单位来重新计算获得。
* React 使用 `Calculator` 组件提供的新 props 分别调用两个 `TemperatureInput` 子组件的 `render` 方法来获取子组件的 UI 呈现。
* React 调用 `BoilingVerdict` 组件的 `render` 方法，并将摄氏温度值以组件 props 方式传入。
* React DOM 根据输入值匹配水是否沸腾，并将结果更新至 DOM。我们刚刚编辑的输入框接收其当前值，另一个输入框内容更新为转换后的温度值。

得益于每次的更新都经历相同的步骤，两个输入框的内容才能始终保持同步。

## 学习小结 {#lessons-learned}

> 在 React 应用中，任何可变数据应当只有一个相对应的唯一“数据源”。通常，state 都是首先添加到需要渲染数据的组件中去。然后，如果其他组件也需要这个 state，那么你可以将它提升至这些组件的最近共同父组件中。你应当依靠 [自上而下的数据流](/docs/state-and-lifecycle.html#the-data-flows-down)，而不是尝试在不同组件间同步 state。

Reviewer 3: ```suggestion
在 React 应用中，任何可变数据应当只有一个相对应的唯一“数据源”。通常，state 都是首先添加到需要渲染数据的组件中去。然后，如果其他组件也需要这个 state，那么你可以将它提升至这些组件的最近共同父组件中。你应当依靠[自上而下的数据流](/docs/state-and-lifecycle.html#the-data-flows-down)，而不是尝试在不同组件间同步 state。
```

让我们来重新梳理一下当你对输入框内容进行编辑时会发生些什么：

* React 在 DOM 的 `<input>` 上调用被指定为是 `onChange` 的函数。在本例中，它是 `TemperatureInput` 组件的 `handleChange` 方法。
* `TemperatureInput` 组件的 `handleChange` 方法使用新输入的值调用 `this.props.onTemperatureChange()`。这个组件的 props 包括 `onTemperatureChange` 在内，都是由它的父组件 `Calculator` 提供的。
* 当 `Calculator` 组件先前渲染后，它用于摄氏度输入的子组件 `TemperatureInput` 中的 `onTemperatureChange` 被指定为了 `Calculator` 组件中的 `handleCelsiusChange` 方法，并且，其用作华氏度输入的子组件 `TemperatureInput` 中的 `onTemperatureChange` 被指定为了 `Calculator` 组件中的 `handleFahrenheitChange` 方法。因此，哪个输入框被编辑将决定 `Calculator` 组件上哪个对应的方法被调用。
* 在这些方法内部，`Calculator` 组件通过使用新的输入值与当前输入框对应的温度计量单位来调用 `this.setState()` 进而请求 React 重新渲染自己本身。
* React 调用 `Calculator` 组件的 `render` 方法得到组件的 UI 呈现。温度转换在这时进行，两个输入框中的数值通过当前输入温度和其计量单位来重新计算获得。
* React 使用 `Calculator` 组件提供的新 props 分别调用两个 `TemperatureInput` 子组件的 `render` 方法来获取子组件的 UI 呈现。
* React 调用 `BoilingVerdict` 组件的 `render` 方法，并将摄氏温度值以组件 props 方式传入。
* React DOM 根据输入值匹配水是否沸腾，并将结果更新至 DOM。
* 使用沸腾判定更新 DOM 并匹配所需的输入值。我们最近编辑的那个输入框接收其当前数值，而另一个输入框内的值更新为经过转换的温度数值。

得益于每次的更新都经历相同的步骤，两个输入框的内容才能始终保持同步。

## 学习小结 {#lessons-learned}

> 在React应用中，任何可变数据应当只有一个相对应的唯一“数据源”。通常，state 都是首先添加到需要渲染数据的组件中去。然后，如果其他组件也需要这个 state，那么你可以将它提升至这些组件的最近共同父组件中。你应当依靠 [自上而下的数据流](/docs/state-and-lifecycle.html#the-data-flows-down)，而不是尝试在不同组件间同步 state。

Author: 这个链接首字符虽然是中文，但是为了与之前保持统一所以空了一个格子

虽然提升 state 方式比双向绑定方式需要编写更多的“样板”代码，但带来的好处是，排查和隔离 bug 所需的工作量将会变少。因为任何“活”在某些组件中的状态数据，仅有该组件自己能够修改它，这样 bug 的排查范围就被大大地被缩减了。此外，你也可以使用自定义逻辑来拒绝或转换用户的输入。

> 如果某些数据可以由 props 或 state 推导得出，那么它就不应该存在于 state 中。举个例子，本例中我们没有将 `celsiusValue` 和 `fahrenheitValue` 一起保存，而是仅保存了最后修改的 `temperature` 和它的 `scale`。这是因为另一个输入框的温度值始终可以通过这两个值以及组件的 `render()` 方法获得。这使得我们能够清除输入框内容，亦或是，对输入框内的数值在不损失任何精度的前提下进行四舍五入的计算。

Author: 是理解有些偏差了，我改一下。这句话的意思是不损失用户输入的那个输入框内数值的精度，而只对另一个经过转换的输入框内数值做四舍五入的操作。
